File: 1398-number-of-ways-to-stay-in-the-same-place-after-some-steps/number-of-ways-to-stay-in-the-same-place-after-some-steps.java
class Solution {
    public int numWays(int steps, int arrLen) {
        int[][] counts = new int[2][arrLen];
        counts[0][0] = 1;
        for (int i = 0; i < steps; i++) {
            for (int j = 0; j < Math.min(arrLen, steps-i); j++) {
                counts[(i+1)%2][j] = counts[i%2][j];
                if (j > 0) 
                    counts[(i+1)%2][j] = (counts[(i+1)%2][j]+counts[i%2][j-1]) % 1000000007;
                if (j < arrLen-1) 
                    counts[(i+1)%2][j] = (counts[(i+1)%2][j]+counts[i%2][j+1]) % 1000000007;                
            }
        }
        return counts[steps%2][0];
    }
}

File: 1398-number-of-ways-to-stay-in-the-same-place-after-some-steps/number-of-ways-to-stay-in-the-same-place-after-some-steps.py
from functools import cache

class Solution:
    def numWays(self, steps: int, arrLen: int) -> int:
        @cache
        def getAnswer(i, step):
            if step == 0 or i < 0:
                return 1 if i == 0 else 0
            answer = getAnswer(i, step-1)
            if i > 0:
                answer += getAnswer(i-1, step-1)
            if i < arrLen-1:
                answer += getAnswer(i+1, step-1)
            return answer % 1000000007
        return getAnswer(0, steps)
        